File: bench/memory/pause.cpp
#include <benchmark/benchmark.h>

#include <atomic>
#include <thread>

std::atomic<bool> lock_flag{false};

static void BM_SpinlockWithPause(benchmark::State& state) {
  for (auto _ : state) {
    for (int i = 0; i < 100'000; ++i) {
      while (lock_flag.exchange(true, std::memory_order_acquire)) {
#if defined(__x86_64__) || defined(_M_X64)
        __builtin_ia32_pause();
#elif defined(__aarch64__)
        asm volatile("yield");
#endif
      }
      lock_flag.store(false, std::memory_order_release);
    }
  }
}

static void BM_SpinlockWithoutPause(benchmark::State& state) {
  for (auto _ : state) {
    for (int i = 0; i < 100'000; ++i) {
      while (lock_flag.exchange(true, std::memory_order_acquire)) {
        // busy spin without pause
      }
      lock_flag.store(false, std::memory_order_release);
    }
  }
}

BENCHMARK(BM_SpinlockWithPause);
BENCHMARK(BM_SpinlockWithoutPause);

BENCHMARK_MAIN();

File: bench/memory/pause.zig
const std = @import("std");
const builtin = @import("builtin");

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

fn spinPause() void {
    if (builtin.cpu.arch == .x86_64) {
        asm volatile ("pause");
    } else if (builtin.cpu.arch == .aarch64) {
        asm volatile ("yield");
    }
}

pub fn main() !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    var lock = std.atomic.Value(bool).init(false);

    // With pause
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            for (0..100_000) |_| {
                while (lock.swap(true, .acquire)) spinPause();
                lock.store(false, .release);
            }
            total += timer.read();
        }
        try printResult("BM_SpinlockWithPause", total, iters);
    }

    // Without pause
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            for (0..100_000) |_| {
                while (lock.swap(true, .acquire)) {}
                lock.store(false, .release);
            }
            total += timer.read();
        }
        try printResult("BM_SpinlockWithoutPause", total, iters);
    }
}
